File: content/snippets/python/design_patterns/visitor.py
from abc import ABC, abstractmethod

class Visitor(ABC):
    @abstractmethod
    def visit_circle(self, circle: "Circle") -> None:
        pass
    
    @abstractmethod
    def visit_rectangle(self, rectangle: "Rectangle") -> None:
        pass

class Element(ABC):
    @abstractmethod
    def accept(self, visitor: Visitor) -> None:
        pass

class Circle(Element):
    def accept(self, visitor: Visitor) -> None:
        visitor.visit_circle(self)
    
    def draw(self) -> None:
        print("Drawing circle")

class Rectangle(Element):
    def accept(self, visitor: Visitor) -> None:
        visitor.visit_rectangle(self)
    
    def draw(self) -> None:
        print("Drawing rectangle")

class DrawVisitor(Visitor):
    def visit_circle(self, circle: Circle) -> None:
        circle.draw()
    
    def visit_rectangle(self, rectangle: Rectangle) -> None:
        rectangle.draw()

circle = Circle()
rectangle = Rectangle()
visitor = DrawVisitor()

circle.accept(visitor)
rectangle.accept(visitor)

File: content/snippets/go/design_patterns/visitor.go
package main

import "fmt"

type Visitor interface {
	VisitCircle(*Circle)
	VisitRectangle(*Rectangle)
}

type Element interface {
	Accept(Visitor)
}

type Circle struct{}

func (c *Circle) Accept(v Visitor) {
	v.VisitCircle(c)
}

func (c *Circle) Draw() {
	fmt.Println("Drawing circle")
}

type Rectangle struct{}

func (r *Rectangle) Accept(v Visitor) {
	v.VisitRectangle(r)
}

func (r *Rectangle) Draw() {
	fmt.Println("Drawing rectangle")
}

type DrawVisitor struct{}

func (dv *DrawVisitor) VisitCircle(c *Circle) {
	c.Draw()
}

func (dv *DrawVisitor) VisitRectangle(r *Rectangle) {
	r.Draw()
}

func main() {
	circle := &Circle{}
	rectangle := &Rectangle{}
	visitor := &DrawVisitor{}

	circle.Accept(visitor)
	rectangle.Accept(visitor)
}
